### DISCUSSION TITLE
Script - Separating bytes from ir4 during i2c reads

### DISCUSSION THREAD

**Discussion by letshin** [arendst/Tasmota#20181] [Q&A] [answered]
Edit: This Q&A was called something else but has been changed as I have made several tests and adjustments.

Hi, I have two i2c devices connected to an ESP 12F on GPIO2 and GPIO14 (SDA and SCL respectively)

The SHT4x works with the preset drivers, I am trying to change to script so I can read from both the SHT4x as well as the AGS10 (datasheet: http://www.aosong.com/userfiles/files/Datasheet%20AGS10.pdf: which appears to be reading 5 bytes from 0x00) using script.

I'm able to read the sensor using a minimal wire code from the Arduino IDE
```
#include "Wire.h"

uint8_t buffer[5];

void setup()
{
 delay(1000);
 Serial.begin(115200);

 Wire.begin(2,14);
 Wire.setClock(15000L);   
}

void loop()
{
 delay(3000);
 Wire.requestFrom(26, 5);
 for ( int i = 0; i < 5; i++)
 {
   buffer[i] = Wire.read();
 }
 Serial.println();

 //  CONVERT RAW DATA
 Serial.print("STAT:\t");
 Serial.println(buffer[0]);
 Serial.print("PPB:\t");
 Serial.println(buffer[1] * 65536UL + buffer[2] * 256 + buffer[3]);
 Serial.print("1:\t");
 Serial.println(buffer[1]);
 Serial.print("2:\t");
 Serial.println(buffer[2]);
 Serial.print("3:\t");
 Serial.println(buffer[3]);      
 Serial.print("CRC:\t");
 Serial.println(buffer[4]);
 Serial.println();
}
```
I'm trying to replicate this in script but am not sure how to 1)Separate out the bytes, and 2) Reduce the i2c speed. Can anyone advice please. Here is where I am so far on my current script. The first byte from 0x00 (the status bit) reads 16 which is the same as when I'm reading it from the code above, but I'm not sure how to extract the other values.
```
>D
found=0
val0=0
val1=0
val2=0
val3=0
val4=0
val5=0
val6=0
>B
; check device
found=ia(0x1a)
>S
if found>0 {
val0=ir1(0x00)
val2=ir4(0x00)

val4=val2-(256*256*256)

print %val0%
print %val1%
print %val2%
print %val3%
print %val4%
print %val5%
print %val6%
}
```
Output:
```
06:24:10.410 16.00
06:24:10.412 0.00
06:24:10.414 16777460.00
06:24:10.416 0.00
06:24:10.418 244.00
06:24:10.420 0.00
06:24:10.422 0.00
```

**Comment by barbudor** [comment]
If you want to drive the SHT4x from Scripting-Language, you need to disable the native driver using `i2cdriver15 0`. Don't mix both.

I'm not sure that scripting language can work with 2 different I2C peripheral at the same time.
Anyhow the address for the AGS10 is 0x34, not 0x44

**Reply by letshin** [reply]
Its actually 0x1A for the AGS10 as confirmed from the datasheet and also from i2cscan. I just didn't list it above.

I want to run SHT4x using the native driver and read the AGS10 using script. WOuld this be doable?

**Reply by barbudor** [reply]
Ah, yes the data sheet write 0x34 for read and 0x35 for write which include the R/W bit. So definitively 0x1A

> I want to run SHT4x using the native driver and read the AGS10 using script. WOuld this be doable?

That should most probably be ok

My suggestion would be to 1st develop the AGS driver without connecting the SHT4x.
So when that would be working you can connect the SHT4x and confirm this all works together

**Reply by letshin** [reply]
Thanks.

I'm trying that now. I was being stupid - using tasmo compiler and did not realise that i2c display means i2c AND display. I thought i2c was enabled by default. Now I can set the i2c sda and scl in teh config so I will try developing it now.

**Reply by letshin** [reply]
I'm struggling to understand the datasheet - does it mean I just have to read 5 bits from 0x00 and convert that to ppb? Since only the bit 2, 3, and 4 contain the data I can just read 0x01, 0x02, and 0x03?

**Reply by sfromis** [reply]
Reading is one whole byte at a time, next step may be to isolate interesting bits from each byte.

**Reply by sfromis** [reply]
If you are referring to this, the intention is that you read 5 bytes, check the CRC, use the low order 4 bits of the first byte, and take the TVOC value from the next 3 bytes. 
![image](https://github.com/arendst/Tasmota/assets/47082390/758771a5-207c-4f45-84eb-9ddbcd03f692)

(As I'm not using scripting, I have no how-to suggestions for doing this)

**Reply by letshin** [reply]
Thanks. If the intention is to read 5 bytes from 0x00, then am I right in saying that reading 0x01, 0x02, and 0x03 should be sufficient to get the TVOC data? This makes no sense to me because
1. Reading a single byte from 0x03 `ir(0x03)` returns 3
2. Reading a single byte from 0x01 and 0x02 `ir(0x01) and ir(0x02)`returns 0 each
3. Reading 3 bytes at 0x01 `ir3(0x01)` returns 32007.00
4. There is no change in any of the bytes when I'm blowing into the sensor
which means I'm totally misunderstanding this :/

**Reply by sfromis** [reply]
Well, ideally you first read all 5 bytes to be able to check if the CRC is correct before starting to give meaning to the values. Ok, not *strictly* required, but that's how the data is intended to be used.

As 32007 is 0x007d07 (depending on endianess), that does not match bytes 0,0 and 3 you found. Not having experience with the device (or doing I2C in scripting) I do not have any specifics, but it may not be a given that you can randomly read whatever byte you like.

**Reply by letshin** [reply]
The chip also requires the i2c speed to not exceed 15khz so this may be why as well. I guess I'll need to wait to see if someone like gemu would have any thoughts on this

**Reply by sfromis** [reply]
If they specifically mean bus speed (and not something like sample rate), 15 KHz would be surprisingly slow, typical I2C bus speeds are standard 100 KHz or fast 400 KHz. Looks like Tasmota is using 100 KHz, or maybe 400 KhZ on ESP32. However, the default expectation is that the bus would just be slowed down, which should be no problem unless the bus is shared with devices needing a higher speed.

**Reply by letshin** [reply]
Yes, I believe this would be bus speed.  As there is a similar chip ags02ma that Adafruit uses (and warns on the clock speeds required) from what I assume to be the same family that is also limited to a low 30khz communication speed (datasheet : https://cdn-shop.adafruit.com/product-files/5593/datasheet+ags02ma.pdf). When I chose this particular chip I kinda didn't realise what that meant when going through the datasheet and am now paying the price. :/ There appears to be away to reduce the i2c rate but I need to spend some time understanding it first (https://github.com/esp8266/Arduino/issues/2524)

**Reply by sfromis** [reply]
What I meant is that I see statements like this on the Internet:
> If multiple devices with different i2c rates are interfaced on same bus, then the lowest should be the speed.

However, I do not know if this will always work, or if there may be caveats. Thus "I'd not be surprised" if it works anyway, but not going to make any assertions either way.

**Reply by letshin** [reply, marked as answer]
Sorted it for both sensors (without slowing down i2c speed, keeping stock values whatever they are) as :

```
>D
found=0
status=0
val0=0
val1=0
topic="test"
hum=0
temp=0

>B
; check device
found=ia(0x1a)

>S
;update TVOC every 3s
if upsecs%3==0 {
 val1=ir4(0x00)
 status=val1>>24
}
if found>0 {
 if val1>(status<<24) {
  val0=val1-(status<<24)
  ;print %val0%
 }
}

>T
temp=SHT4X#Temperature
hum=SHT4X#Humidity
=>Publish tele/%topic%/SENSOR {"Sensor":{"TVOC":%val0%,"Temperature":%temp%,"Humidity":%hum%}}
```
